mixer/adapter/stackdriver/metric/bufferedClient.go

```diff
@@ -75,12 +89,18 @@ func (b *buffered) Record(toSend []*monitoringpb.TimeSeries) {
 }
 
 func (b *buffered) Send() {
+	// Retry timeseries in retry buffer. Retry before normal push so that retried points could be
+	// preserved thus view would be more accurate and order is guaranteed.
+	// If timeseries is again failed to create, merge it with newer timeseries in normal buffer.
+	b.retry()
```

Comment:
It looks like you're only retrying as part of Send. That means you're still going to just send a big burst of metrics once/minute (or whatever the frequency Send is called at). Ideally the writes would be spread out over the course of the minute.

I asked the Monitoring API team about retry and rate limiting and here were their suggestions:

"standard retry practices are only on 503/504, (b) maximum number of retries at about 3, with exponential backoff"
"the other thing we advise is a priority queue for writes that brings retries to the front" (this seems like what you've got here)

Is it okay for Send() to block? If so, here is the logic I suggest:

- Put retryBuffer + new points in a single slice of points to write
- Partition into batches with batchTimeSeries
- Do logic like this to send the batches:
t := time.NewTicker(cfg.PushInterval/len(batches)
defer t.Stop()
var retry []retryTimeSeries
for _, b := range batches {
  unsent := pushMetrics(b)
  retry = append(retry, unsent...)
  <-t.C
}
b.retryBuffer = retry

This will smear the writes across the push interval, giving you the best chance of not hitting rate limits. It's not quite exponential backoff, but it's still limited backoff. I'll comment below on the right error handling.

BTW, the max batch size is 200 points. It looks like you currently are using 200 in prod but 100 in the test case?

Reply:
Do we really need to spread push calls? Monitoring API's quota is 6000 per minute.. I don't expect it could be reached... Any other reasons why this would be meaningful?

batch size limit is passed into buffer client when it is initiated. 100 is just for test.

Reply:
Yes, the Monitoring API team requested that you spread writes out across time if at all possible. It's a thundering herd problem.

Reply:
Done.